lib/spack/spack/test/patch.py

```diff
@@ -44,7 +44,7 @@ def mock_apply(monkeypatch):
     def check_expand(self, *args, **kwargs):
         # Check tarball expansion
         if spack.util.compression.allowed_archive(self.url):
-            file = os.path.join(self.path, 'foo.txt')
```

Comment:
@mathstuf I think the test is correct as it is, even though the parts involved would surely benefit from a major refactoring. The expression:
```python
file = os.path.join(self.path, 'foo.txt')
```
should contain already the part `spack-expanded-archive`. This is due to how [stage set its path](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/stage.py#L219):
```python
# Try to construct here a temporary name for the stage directory
# If this is a named stage, then construct a named path.
if path is not None:
    self.path = path
else:
    self.path = join_path(spack.stage_path, self.name)
# NOTE: here type(self) == Stage and the path DOES NOT contain `spack-expanded-archive`
```
plus how `Patch.path` [builds](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/patch.py#L138) on that:
```python
# Here type(self) is UrlPatch
self.path = os.path.abspath(
    os.listdir(patch_stage.path).pop()
)
```
Due to `os.listdir` we have the directory already included.

Reply:
Hmm, it fails as-is for me and needed this fix. The path had been coming out to `/path/to/tar.gz/foo.txt`.

Reply:
On which platform are you working? I think we may be hitting  platform dependent behavior then.

Reply:
Fedora 25.